### PR TITLE
feat: add Available status to Kepler

### PR DESCRIPTION
**Pull request by sthaha**
This PR adds the Availability condition to the Kepler resource status to
 indicate whether the kepler deployment is available for use. It also ensures
 that the resources are properly deleted when Kepler is deleted. 

 Additionally, `kubectl get kepler` now prints the following

 ```
NAME      PORT   DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE-SELECTOR
invalid   8122   0         0         0                                {"kubernetes.io/os":"linux"}
kepler    9103   3         3         3       3            3           {"kubernetes.io/os":"linux"}
```
Also fixes 
* https://github.com/sustainable-computing-io/kepler-operator/issues/179

### CODE REVIEW

File: pkg/controllers/kepler.go [hunk 1]
```diff
@@ -182,31 +193,95 @@ func (r KeplerReconciler) updateStatus(ctx context.Context, req ctrl.Request, re
...
+
+	status := ds.Status
+
+	// NumberUnavailableThe number of nodes that should be running the daemon pod
+	// and have none of the daemon pod running and available (ready for at least
+	// spec.minReadySeconds)
+	if x := status.NumberUnavailable; x > 0 {
+		available.Status = v1alpha1.ConditionFalse
+		available.Reason = v1alpha1.DaemonSetUnavailable
+		available.Message = fmt.Sprintf("Kepler pod is not running and available in %d nodes", x)
```
Comment (vprashar2929) [inline pkg/controllers/kepler.go:234]: Shouldn't the message be `Kepler pod is not running and not available`?

File: pkg/controllers/kepler.go [hunk 2]
```diff
@@ -167,13 +170,21 @@ func (r KeplerReconciler) updateStatus(ctx context.Context, req ctrl.Request, re
 			r.logger.V(6).Info("Reconcile has deleted kepler; skipping update")
 			return nil
 		}
+		now := metav1.Now()
 
 		if recErr != nil {
```
Comment (sthaha) [inline pkg/controllers/kepler.go:175]: What do you feel about adding the Daemonset status - NumberAvailable, Schedueld, etc here and modify the Kepler CRD to show readiness just like a pod? 
```
oc get keplers 

NAME                 DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE                           kepler                            3         3         3       3            3           kubernetes.io/os=linux   41
```

Comment (sthaha) [reply pkg/controllers/kepler.go:175]:
```
 ❯ kubectl get keplers

NAME     DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODESELECTOR
kepler   3         3         3       3            3           {"kubernetes.io/os":"linux"}
        ────────────────────────────────────────────

```
This is how it looks after the implementation

File: pkg/controllers/kepler.go [hunk 3]
```diff
@@ -168,45 +171,132 @@ func (r KeplerReconciler) updateStatus(ctx context.Context, req ctrl.Request, re
...
+	// NumberAvailable: The number of nodes that should be running the daemon pod
+	// and have one or more of the daemon pod running and available (ready for at
+	// least spec.minReadySeconds)
+	if ds.NumberAvailable < ds.DesiredNumberScheduled {
+		c.Status = v1alpha1.ConditionUnknown
+		c.Reason = v1alpha1.DaemonSetPartiallyAvailable
+		c.Message = fmt.Sprintf("Only %d of total %d pods are running and available; pending: %d",
+			ds.NumberAvailable, ds.DesiredNumberScheduled, ds.NumberUnavailable)
+		return c
+	}
```
Comment (vimalk78) [inline pkg/controllers/kepler.go:262]: will this block ever hit?

File: pkg/utils/test/assertions.go [hunk 4]
```diff
@@ -69,7 +69,15 @@ func assertOption(fns ...AssertOptionFn) AssertOption {
 	return option
 }
 
-func (f Framework) AssertResourceExits(name, ns string, obj client.Object, fns ...AssertOptionFn) {
+func (f Framework) WaitUntil(msg string, fn wait.ConditionFunc, fns ...AssertOptionFn) {
```
Comment (vimalk78) [inline pkg/utils/test/assertions.go:72]: rename this function to hint that there is an assert inside?

File: bundle/manifests/kepler-operator.clusterserviceversion.yaml [hunk 5]
```diff
@@ -25,7 +25,7 @@ metadata:
     capabilities: Basic Install
     categories: Monitoring
     containerImage: quay.io/sustainable_computing_io/kepler-operator:0.6.2
-    createdAt: "2023-08-28T09:36:02Z"
+    createdAt: "2023-08-30T03:55:14Z"
```
Comment (vimalk78) [inline bundle/manifests/kepler-operator.clusterserviceversion.yaml:28]: should this date change with every PR? or change only from release Job in CI?

File: pkg/controllers/kepler.go (hunk 2 above)
Comment (vimalk78) [reply pkg/controllers/kepler.go]: will this be printed every time we get `kepler` or need to pass a flag to `oc` ?

File: bundle/manifests/kepler.system.sustainable.computing.io_keplers.yaml [hunk 6]
```diff
@@ -14,7 +14,29 @@ spec:
...
+    - jsonPath: .status.numberReady
+      name: Ready
+      type: integer
+    - jsonPath: .status.updatedNumberScheduled
+      name: Up-to-date
+      type: integer
+    - jsonPath: .status.numberAvailable
+      name: Available
+      type: integer
+    - jsonPath: .spec.exporter.nodeSelector
```
Comment (vimalk78) [inline bundle/manifests/kepler.system.sustainable.computing.io_keplers.yaml:36]: print Tolerations also?

File: pkg/reconciler/deleter.go [hunk 7]
```diff
@@ -19,24 +19,55 @@ package reconciler
...
+	if a > b {
+		return a
+	}
+	return b
 }
 
 func (r Deleter) Reconcile(ctx context.Context, c client.Client, scheme *runtime.Scheme) Result {
+	objKey := client.ObjectKeyFromObject(r.Resource)
+
+	fmt.Printf("%s: Delete %v \n", time.Now().UTC().Format(time.RFC3339), objKey)
```
[the remark this answers is not in the document]
Comment (sthaha) [reply pkg/reconciler/deleter.go:48]: Thank you!

File: pkg/utils/test/assertions.go (hunk 4 above)
Comment (sthaha) [reply pkg/utils/test/assertions.go:72]: Any suggestion?

File: pkg/components/exporter/exporter.go [hunk 8]
```diff
@@ -257,7 +267,7 @@ func NewClusterRoleBinding(c components.Detail) *rbacv1.ClusterRoleBinding {
 				Kind:       "ClusterRoleBinding",
 			},
 			ObjectMeta: metav1.ObjectMeta{
-				Name:   ClusterRoleName,
+				Name:   ClusterRoleBindingName,
```
[the remark this answers is not in the document]
Comment (sthaha) [reply pkg/components/exporter/exporter.go:270]: Thanks to the printfs :D

File: pkg/api/v1alpha1/kepler_types.go [hunk 9]
```diff
@@ -180,6 +188,36 @@ type Condition struct {
...
+	// The number of nodes that should be running the kepler pod and have one or
+	// more of the kepler pod running and available
+	// +optional
+	NumberAvailable int32 `json:"numberAvailable,omitempty"`
+
+	// The number of nodes that should be running the
+	// kepler pod and have none of the kepler pod running and available
+	// +optional
+	NumberUnavailable int32 `json:"numberUnavailable,omitempty"`
+
```
Comment (vimalk78) [inline pkg/api/v1alpha1/kepler_types.go:220]: can we somehow reduce these ? some of these may be transient. e.g. should we keep both `numberAvailable` and `numberReady` ?

File: pkg/controllers/kepler.go (hunk 3 above)
Comment (sthaha) [reply pkg/controllers/kepler.go:262]: Good question, I thought this should be a state that happens .. Searching for this got me 
https://github.com/kubernetes/kubectl/blob/master/pkg/polymorphichelpers/rollout_status.go#L108C5-L108C5

File: pkg/controllers/kepler.go (hunk 2 above)
Comment (sthaha) [reply pkg/controllers/kepler.go]: No, that will be the default after this patch merges, due to `printcolumn` that is added.

**Comment by vimalk78** [comment]
add some PR description?

File: bundle/manifests/kepler-operator.clusterserviceversion.yaml (hunk 5 above)
Comment (sthaha) [reply bundle/manifests/kepler-operator.clusterserviceversion.yaml:28]: it is okay to change this whenever we run bundle update.

File: bundle/manifests/kepler.system.sustainable.computing.io_keplers.yaml (hunk 6 above)
Comment (sthaha) [reply bundle/manifests/kepler.system.sustainable.computing.io_keplers.yaml:36]: I didn't see tolerations being print as part of kc get daemonset ... I am open to adding it though.

File: pkg/api/v1alpha1/kepler_types.go (hunk 9 above)
Comment (sthaha) [reply pkg/api/v1alpha1/kepler_types.go:220]: I think we should maintian 1-1 mapping to the daemonset.

File: bundle/manifests/kepler.system.sustainable.computing.io_keplers.yaml (hunk 6 above)
Comment (sthaha) [reply bundle/manifests/kepler.system.sustainable.computing.io_keplers.yaml:36]: I had some trouble adding toleration, default. I will update that in a separate PR.

Comment (sthaha) [reply bundle/manifests/kepler.system.sustainable.computing.io_keplers.yaml:36]: Created - https://github.com/sustainable-computing-io/kepler-operator/issues/186

File: pkg/controllers/kepler.go [hunk 10]
```diff
@@ -182,31 +193,95 @@ func (r KeplerReconciler) updateStatus(ctx context.Context, req ctrl.Request, re
...
+		ObservedGeneration: k.Generation,
+		LastTransitionTime: metav1.Now(),
+	}
+
+	// get daemonset owned by kepler
+	ds := appsv1.DaemonSet{}
+	key := types.NamespacedName{Name: exporter.DaemonSetName, Namespace: components.Namespace}
+	if err := r.Client.Get(ctx, key, &ds); err != nil {
+		available.Status = v1alpha1.ConditionFalse
+		available.Reason = v1alpha1.DaemonSetNotFound
```
[the remark this answers is not in the document]
Comment (sthaha) [reply pkg/controllers/kepler.go]: Yes, good idea and done!

File: pkg/utils/test/assertions.go (hunk 4 above)
Comment (sthaha) [reply pkg/utils/test/assertions.go:72]: @vimalk78 , I am going to merge this as is, we can rename this test utility when we have a better idea :)

File: bundle/manifests/kepler.system.sustainable.computing.io_keplers.yaml (hunk 6 above)
Comment (sthaha) [reply bundle/manifests/kepler.system.sustainable.computing.io_keplers.yaml:36]: @vimalk78 , I have added tolerations to be print but the default one isn't listed (yet) which the issue above should fix.

File: pkg/utils/test/assertions.go (hunk 4 above)
Comment (vimalk78) [reply pkg/utils/test/assertions.go:72]: I missed seeing the the function signature, which has `AssertOptionFn` so there should be no ambiguity.

**Review by vimalk78 — APPROVED** [review]
lgtm

**Pull request merged**